### DISCUSSION TITLE
Errors and Tutorials

### DISCUSSION THREAD

**Discussion by DeadTomGC** [SpiceSharp/SpiceSharpBehavioral#53] [Ideas]
Hey Folks!

TLDR: I'm planning to write a more detailed tutorial for mid-simulation circuit updates.  Also, I'm wondering if it makes sense to add basic Invalid Input checking on things like DcValue's for Voltage Sources.  Also, SpiceSharp/Behavioral is fantastic. 

I've been working with SpiceSharp and SpiceSharpBehavioral for a couple months now with great results!
I'm simulating pretty intense circuits with many orders of magnitude in them that professional grade spice programs struggle with sometimes, and I haven't had a REAL issue yet.  

The basic use case for what I'm doing is simulating hardware that doesn't yet exist, and then seeing how a control algorithm reacts to that simulated hardware.  So, mid-simulation, the control hardware/software is polled for an updated set of control signals, and then these new signals are injected into the simulation using voltage sources by setting their "DcValue" in the "BeforeLoad" callback.  

Along the way I had many issues, though!  Most were from not understanding how spicesharp worked, or just having bugs I didn't quite know how to trace down.  

I think a more extensive tutorial on doing mid-simulation updates to components would help one of these issues, and I think a little more error checking could help the other.

The tutorial would be one I'd be happy to write since it's born from my total ignorance of how spice simulations work.  So it'd be written for dummies like me. So, that's going on my TODO list.

The error checking though I'm unsure about... the principle issue I ran into was that I was setting a Voltage Source DC value to a NaN (divide by 0 error).  The error I got though was very generic though.  I feel like software should be smart enough to identify when the USER screws up his/her input as compared to the simulation actually having an issue. (The usual,  "invalid input" sort of error)

I'm really happy about this tool, and I'm sure I haven't even realized what its full potential is.  It's making a big impact in how software is developed and tested.

**Comment by svenboulanger** [comment]
Hi!

Thanks for the message, always good to have some feedback :-) Happy to see it is suiting your needs (for the most part). And nice use-case by the way 👍  This is the kind of stuff I was thinking of when building Spice#. If you have things you would like to see differently, feel free to post something on the [Spice# issue page](https://github.com/SpiceSharp/SpiceSharp/issues).

The [documentation](https://spicesharp.github.io/SpiceSharp/) having changing parameters mid-simulation can indeed use some more work. I originally added it as one of the use cases, but I didn't have much feedback on it so it didn't get more attention as Spice# got updated. Documentation is also very time-consuming for me so hard to keep updated.

Having some kind of additional invalid input checking does make sense to me. This will become a noted issue for the core Spice# package, and I expect it will be part of the upcoming Spice# v3.1. But first I need to work on the new C# source generator which will be very useful for making custom components as well.

Cheers!

**Reply by DeadTomGC** [reply]
I'll have to update my version to latest and see if it fixes an error I had "Illegal one-byte branch at position: 65. Requested branch was: 354"  It sounds like that update could be related.  My work around was to just VERY dumbly cut all the _S's from the OpCodes.

**Comment by DeadTomGC** [comment]
First Draft to get main gist across:

### Basics of Using SpiceSharp for Hardware/Software in the Loop Testing ###
See [This Page](https://spicesharp.github.io/SpiceSharp/articles/tutorials/changing_parameters/changing_parameters_during_simulation.html) for the basics of modifying components mid simulation.

Here is the basic flow for this kind of simulation:
![SimulationFlow](https://user-images.githubusercontent.com/1928571/118747814-546fa800-b820-11eb-82ab-5a016a9ca8b9.png)

The important take-away here is that `BeforeLoad` gets called many times while attempting to make the simulation converge.  This means that any changes being made to the components in the simulation in `BeforeLoad` MUST look continuous with respect to time. (No Stair-Stepping)
`double time = tran.GetState<IIntegrationMethod>().Time;`
While the simulation is searching for an acceptable timestep, it starts with a larger timestep, and then works backwards from there.  Do NOT assume time will always march forward from the perspective of the `BeforeLoad` callback.

Also, the data from a timestep is only real once it is in the `ExportSimulationData` Callback.  Take this into account when choosing when to query your external HW/SW control system.

Example code to get DC Voltage Source Component Handle in `AfterSetup`:
`SpiceSharp.Components.CommonBehaviors.IndependentSourceParameters V1sp = null;`
`var eb = transientSimulation.EntityBehaviors[sourceName];`
` eb.TryGetParameterSet(out DCSupplysp);`

Example code to modify the DC Voltage Source in `BeforeLoad`:
`double time = tran.GetState<IIntegrationMethod>().Time;`
`DCSupplysp.DcValue = 0.01*time;`

I'd like to provide a fully working example here, but because of restrictions on what I can upload, it'll take some approval to append that.

This anyway is what I would have found most useful when learning and implementing a system.

**Reply by svenboulanger** [reply]
That looks right to me, so nicely done 🙂
The docs are generated using docfx. You should be able to add documentation by adding and editing markdown and yml files. Just make sure to pr to development if you do. We can look to running docfx then. 
After it gets merged with master, the GitHub pages are updated automatically.
